### PR TITLE
node is able to run klaos local network

### PR DESCRIPTION
**Pull request by asiniscalchi**
## **Type**
enhancement, tests

___

## **Description**
- Added chain spec for `klaos` with development and local testnet configurations.
- Refactored `laos` chain spec to align with `klaos` and use shared configurations.
- Centralized chain spec configuration in `mod.rs` for better code reuse.
- Added support for loading `klaos` specific chain specs in the node command.
- Extended primitives for parachain with new types and used shared primitives in `klaos` and `laos` runtimes.
- Added CI jobs for testing runtime upgrades and running end-to-end tests on the `klaos` network.
- Updated node dependencies to include `klaos-ownership-runtime` and shared primitives.
- Added a new plain JSON spec for the `Caladan` network and a Zombienet configuration for `klaos`.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>9 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>klaos.rs</strong><dd><code>Add Chain Spec for Klaos</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/node/src/chain_spec/klaos.rs
<li>Added a new chain spec file for <code>klaos</code> with development and local <br>testnet configurations.<br> <li> Defined session keys, balances, parachain info, and EVM accounts for <br><code>klaos</code>.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/383/files#diff-4708209aac5bd94fe0970ebb7decc10178e3a4a3dc02bdfd03dad3a5e439b10b">+228/-0</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>laos.rs</strong><dd><code>Refactor Laos Chain Spec</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/node/src/chain_spec/laos.rs
<li>Updated the <code>laos</code> chain spec to align with the new structure used in <br><code>klaos</code>.<br> <li> Simplified the code by removing redundant functions and using shared <br>constants.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/383/files#diff-d37bf7c1411ae01eb45345be8883325c30549c8c4652b6eaedd6692755cbea3a">+10/-69</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>mod.rs</strong><dd><code>Centralize Chain Spec Configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/node/src/chain_spec/mod.rs
<li>Introduced shared constants and helper functions for chain specs.<br> <li> Centralized the configuration for endowed accounts and collator keys.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/383/files#diff-0d26b66fe1eb0a77184d37983feab81059fbf61f3950e47c3cbccb20fee13bf3">+61/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>command.rs</strong><dd><code>Support Klaos Chain Spec Loading</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/node/src/command.rs
<li>Added support for loading <code>klaos</code> specific chain specs.<br> <li> Refactored the spec loading logic to support multiple parachains.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/383/files#diff-58d0894ee6f123313b5d41c83a161d38f23d752263f34cb37498be9b830ef372">+33/-12</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>lib.rs</strong><dd><code>Extend Primitives for Parachain</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/primitives/src/lib.rs
<li>Added <code>AuraId</code> type alias for parachain authority ID.<br> <li> Introduced new types for transaction index and moment.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/383/files#diff-39aed6c4e121e4d540aa051ee063c1733b1f975c761d9b8bd206960284a55ee4">+6/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>lib.rs</strong><dd><code>Refactor Klaos Runtime with Shared Primitives</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/runtime/klaos/src/lib.rs
<li>Refactored to use shared primitives from <br><code>ownership-parachain-primitives</code>.<br> <li> Removed redundant type aliases and constants.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/383/files#diff-e62312042f1e66bd43b36250c910f0563dfd68983ec83173c335a3cd24f43121">+3/-39</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>lib.rs</strong><dd><code>Update Laos Runtime with Shared Primitives</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/runtime/laos/src/lib.rs
<li>Updated to use shared primitives for basic types.<br> <li> Cleaned up unused type aliases and constants.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/383/files#diff-e71ce79c6361a064296c98f1244126573a4ae185adc40f1afca2fc27e5e74c38">+3/-23</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Cargo.toml</strong><dd><code>Update Node Dependencies for Klaos</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/node/Cargo.toml
<li>Added <code>klaos-ownership-runtime</code> as a dependency.<br> <li> Included <code>ownership-parachain-primitives</code> for shared types.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/383/files#diff-a95acab9d237247f78755bd16f676aebbc0591ffb4200abbadb951bd28d18be6">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Cargo.toml</strong><dd><code>Add Aura Consensus Dependency</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/primitives/Cargo.toml
- Included `sp-consensus-aura` for Aura consensus support.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/383/files#diff-1254592a14dcc0a9783c990549ceaa96c04ee38332f4e8e8f1c869f8521592c9">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>build.yml</strong><dd><code>Add CI Jobs for Klaos Network</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/workflows/build.yml
<li>Added a new job for <code>try-runtime-klaos</code> to test runtime upgrades.<br> <li> Introduced <code>e2e-tests-klaos</code> job for running end-to-end tests on the <br><code>klaos</code> network.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/383/files#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721">+60/-3</a>&nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>caladan.plain.json</strong><dd><code>Add Caladan Network Spec</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/specs/caladan.plain.json
- Added a new plain JSON spec for the `Caladan` network.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/383/files#diff-723f55c73e36e8c2129d1cb8e38b6983344fa69487cb03bf53dff938c933c4ab">+127/-1</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>native.klaos.toml</strong><dd><code>Add Zombienet Config for Klaos</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/zombienet/native.klaos.toml
<li>Introduced a new Zombienet configuration for testing the <code>klaos</code> <br>network.

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/383/files#diff-c01f45768a75d83e670915d8d76e4daa3417dae2576dfb6fd2bbc86a318319b4">+47/-0</a>&nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: ownership-chain/node/src/chain_spec/mod.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+use sc_chain_spec::{ChainSpecExtension, ChainSpecGroup};
+use serde::{Deserialize, Serialize};
+
+pub mod klaos;
+pub mod laos;
+
+/// Specialized `ChainSpec` for the normal parachain runtime.
+pub type ChainSpec =
+	sc_service::GenericChainSpec<laos_ownership_runtime::RuntimeGenesisConfig, Extensions>;
```
Comment (asiniscalchi) [inline ownership-chain/node/src/chain_spec/mod.rs:9]: this is used to load chain_spec from file but it has a generic type from laos_ownership_runtime ... is there any way to remove it and make generic ?

File: ownership-chain/node/src/chain_spec/klaos.rs [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+use fp_evm::GenesisAccount;
+use hex_literal::hex;
+use klaos_ownership_runtime::{AccountId, AuraId, Precompiles, REVERT_BYTECODE};
+use sc_service::ChainType;
+use sp_core::{Pair, Public, H160, U256};
+use sp_runtime::traits::Zero;
+use std::{collections::BTreeMap, str::FromStr};
+
+/// List of endowed accounts.
+fn endowed_accounts() -> Vec<AccountId> {
```
Comment (dastanfv) [inline ownership-chain/node/src/chain_spec/klaos.rs:12]: can you move this function and all the other shared types to `mod.rs`?

File: ownership-chain/node/src/chain_spec/klaos.rs [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+		hex!("C0F0f4ab324C46e55D02D0033343B4Be8A55532d").into(),
+	]
+}
+
+/// Specialized `ChainSpec` for the normal parachain runtime.
+pub type ChainSpec =
+	sc_service::GenericChainSpec<klaos_ownership_runtime::RuntimeGenesisConfig, Extensions>;
+
+/// The default XCM version to set in genesis config.
+const SAFE_XCM_VERSION: u32 = staging_xcm::prelude::XCM_VERSION;
```
Comment (dastanfv) [inline ownership-chain/node/src/chain_spec/klaos.rs:34]: this one as well

File: ownership-chain/node/src/chain_spec/klaos.rs [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+/// Helper function to generate a crypto pair from seed
+pub fn get_from_seed<TPublic: Public>(seed: &str) -> <TPublic::Pair as Pair>::Public {
```
Comment (dastanfv) [inline ownership-chain/node/src/chain_spec/klaos.rs:37]: same here and already generic

File: ownership-chain/node/src/chain_spec/klaos.rs [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+	TPublic::Pair::from_string(&format!("//{seed}"), None)
+		.expect("static values are valid; qed")
+		.public()
+}
+
+/// Generate collator keys from seed.
+///
+/// This function's return type must always match the session keys of the chain in tuple format.
+pub fn get_collator_keys_from_seed(seed: &str) -> AuraId {
```
Comment (dastanfv) [inline ownership-chain/node/src/chain_spec/klaos.rs:46]: same

File: ownership-chain/node/src/chain_spec/klaos.rs (hunk 2 above)
Comment (dastanfv) [reply ownership-chain/node/src/chain_spec/klaos.rs:12]: needs `fn endowed_accounts<T: From<[u8;20]>>() -> Vec<T> {}`

**Review by dastanfv — COMMENTED** [review]
shared functions/types should be moved to `mod.rs`

**Review by dastanfv — APPROVED** [review]
lgtm

**Review by ccubu — APPROVED** [review]

**Pull request merged**